clang/lib/Headers/cheriintrin.h

```
@@ -0,0 +1,132 @@
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef __CHERIINTRIN_H
+#define __CHERIINTRIN_H
+
+#if !__has_feature(capabilities)
+#error "<cheriintrin.h> should only be included when CHERI is supported"
+#endif
+
+/*
+ * Provide shorter names for CHERI builtins that are portable across all
+ * CHERI-enabled architectures:
+ */
+#define cheri_getaddress(x) __builtin_cheri_address_get(x)
+#define cheri_setaddress(x, y) __builtin_cheri_address_set((x), (y))
+#define cheri_getbase(x) __builtin_cheri_base_get(x)
+#define cheri_getlength(x) __builtin_cheri_length_get(x)
+#define cheri_getoffset(x) __builtin_cheri_offset_get(x)
+#define cheri_incoffset(x, y) __builtin_cheri_offset_increment((x), (y))
+#define cheri_setoffset(x, y) __builtin_cheri_offset_set((x), (y))
+#define cheri_gettag(x) __builtin_cheri_tag_get(x)
+#define cheri_cleartag(x) __builtin_cheri_tag_clear(x)
+#define cheri_is_subset(x, y) __builtin_cheri_subset_test((x), (y))
+
+/* Bounds setting intrinsics: */
+#define cheri_round_representable_length(x)                                    \
+  __builtin_cheri_round_representable_length(x)
+#define cheri_representable_alignment_mask(x)                                  \
+  __builtin_cheri_representable_alignment_mask(x
+#define cheri_setbounds(x, y) __builtin_cheri_bounds_set((x), (y))
+#define cheri_setbounds_exact(x, y) __builtin_cheri_bounds_set_exact((x), (y))
+
+/* Object types, sealing and unsealing: */
+typedef long cheri_otype_t;
+#if defined(__mips__) || defined(__riscv)
+/* CHERI-MIPS and CHERI-RISCV use negative numbers for hardware-interpreted
+ * otypes */
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)-1)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)-2)
+#else
```

`#elif defined(__morello__)` ?

---

I don't think such a define exists, `defined(__aarch64__)` should work though.

---

This is still unresolved. I think we want to error out (or not define them and error if you try to use them) for unknown architectures.

---

We already error for `!__has_feature(capabilities)`, but sure I can add an `#else #error`